src/Testcontainers.Milvus/MilvusBuilder.cs

```diff
@@ -0,0 +1,75 @@
+namespace Testcontainers.Milvus;
+
+[PublicAPI]
+public class MilvusBuilder : ContainerBuilder<MilvusBuilder, MilvusContainer, MilvusConfiguration>
+{
+    public const string MilvusImage = "milvusdb/milvus:v2.3.10";
+    public const ushort MilvusGrpcPort = 19530;
+    public const ushort MilvusManagementPort = 9091;
+
+    private string? _etcdEndpoint;
+
+    public MilvusBuilder() : this(new MilvusConfiguration())
+        => DockerResourceConfiguration = Init().DockerResourceConfiguration;
+
+    private MilvusBuilder(MilvusConfiguration dockerResourceConfiguration) : base(dockerResourceConfiguration)
+        => DockerResourceConfiguration = dockerResourceConfiguration;
+
+    public override MilvusContainer Build()
+    {
+        Validate();
+        return new MilvusContainer(DockerResourceConfiguration, TestcontainersSettings.Logger);
+    }
+
+    public MilvusBuilder WithEtcdEndpoint(string? etcdEndpoint)
+    {
+        _etcdEndpoint = etcdEndpoint;
+        return this;
+    }
+
+    protected override MilvusBuilder Init()
+    {
+        var builder = base.Init()
+            .WithImage(MilvusImage)
+            .WithEnvironment("COMMON_STORAGETYPE", "local")
+            .WithCommand("milvus", "run", "standalone")
+            .WithPortBinding(MilvusGrpcPort, true)
+            .WithPortBinding(MilvusManagementPort, true)
+            .WithWaitStrategy(
+                Wait.ForUnixContainer()
+                    .UntilHttpRequestIsSucceeded(h => h
+                        .ForPort(MilvusManagementPort)
+                        .ForPath("/healthz")));
+
+        if (_etcdEndpoint is null)
+        {
+            const string etcdYaml = """
+                                    listen-client-urls: http://0.0.0.0:2379
+                                    advertise-client-urls: http://0.0.0.0:2379
+                                    """;
+
+            builder = builder
+                .WithResourceMapping(Encoding.UTF8.GetBytes(etcdYaml), "/milvus/configs/embedEtcd.yaml")
+                .WithEnvironment("ETCD_USE_EMBED", "true")
+                .WithEnvironment("ETCD_DATA_DIR", "/var/lib/milvus/etcd")
+                .WithEnvironment("ETCD_CONFIG_PATH", "/milvus/configs/embedEtcd.yaml");
+        }
+        else
+        {
+            throw new NotImplementedException();
+        }
```

Comment:
Hi, I do not understand this part. `_etcdEndpoint` will always be `null` unless you store it in the container configuration object. Both tests use the variation that includes the `embedEtcd.yaml` file.

The builder API is immutable to support certain configurations for A/B testing. The library creates a new instance of the builder; hence, you cannot keep the reference to the field without storing it somewhere else.

I would expect (propose) the following default configuration:

```csharp
protected override MilvusBuilder Init()
{
    return base.Init()
        .WithImage(MilvusImage)
        .WithPortBinding(MilvusManagementPort, true)
        .WithPortBinding(MilvusGrpcPort, true)
        .WithCommand("milvus", "run", "standalone")
        .WithEnvironment("COMMON_STORAGETYPE", "local")
        .WithEnvironment("ETCD_CONFIG_PATH", EtcdConfigFilePath)
        .WithEnvironment("ETCD_DATA_DIR", "/var/lib/milvus/etcd")
        .WithEnvironment("ETCD_USE_EMBED", "true")
        .WithResourceMapping(DefaultEtcdConfig, EtcdConfigFilePath)
        .WithWaitStrategy(Wait.ForUnixContainer().UntilHttpRequestIsSucceeded(request =>
            request.ForPort(MilvusManagementPort).ForPath("/healthz")));
}
```

Then override the default configuration from the other available public builder APIs, that is how we do it in other modules (_there might be other (better) ways to pass the endpoint to the container, but I think you know better_):

```csharp
public MilvusBuilder WithEtcdEndpoint(string etcdEndpoint)
{
    return WithEnvironment("ETCD_USE_EMBED", "false")
        .WithEnvironment("ETCD_ENDPOINTS", etcdEndpoint);
}
```

This will run both variations. One that uses the embedded etcd and one that uses the sidecar container `quay.io/coreos/etcd`. I think that is what you are trying to achieve?

If you need further guidance or help, just let me know. Happy to assist.

Reply:
Yeah, apologies, the part with an external etcd indeed wasn't finished.

I've made changes following your suggestion - that all makes sense to me. It's slightly odd to define the "embedded" settings as a default and then to override them when WithEtcdEndpoint is called - ideally the embedded settings wouldn't be set at all. For example, as the code is after my change, the embedEtcd.yaml file is always mapped in the container, though it's really needed only for the embedded config (and I'm not sure there's a way to unset environment variables and unmap a resource).

But that's not very important - I'm definitely happy with things as they are if you are. Re a better way for passing the endpoint to the container, I'm not expert here and am mainly following the lead of the [Java module](https://github.com/testcontainers/testcontainers-java/blob/main/modules/milvus/src/main/java/org/testcontainers/milvus/MilvusContainer.java) (we can always improve this in the future etc.)

Reply:
> Yeah, apologies, the part with an external etcd indeed wasn't finished.

No worries. That is what reviews are for.

> or example, as the code is after my change, the embedEtcd.yaml file is always mapped in the container, though it's really needed only for the embedded config (and I'm not sure there's a way to unset environment variables and unmap a resource).

Indeed. This is a bit tricky. Overriding environment variables works fine, but unmapping resources does not. Definitely something I would like to look at and support in the future. Some configurations cannot be unset.

> we can always improve this in the future etc.

I agree. In the future, we can provide, e.g. the sidecar container together with the module, something we are already doing in other modules like [FFmpeg](https://github.com/testcontainers/testcontainers-dotnet/blob/2131fbd27ef394e602b454793216136440584393/src/Testcontainers.WebDriver/WebDriverBuilder.cs#L74-L84). Anyway, the module configuration looks good. Thanks for the adjustments.